### PR TITLE
feat(openebs-upgrade): add taskconfig schema in upgraderesult cr.

### PR DESCRIPTION
**Pull request by imazik**
`Config` is configuration for task executor i.e. a kubernetes job. 

Task-executor gets a list of resources to be upgraded, castemplate name to be used for upgrade and upgrade related config options via a kubernetes configmap. 

_NOTE: There will be one upgrade-result for each unit of upgrade i.e. per upgraded resource._

This PR contains above `Config` schema and logic associated to this schema.

Task executor/job's ConfigMap will be look like:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: task-executor-config
  namespace: default
data:
  upgrade: |
    casTemplate: upgrade-0.8.0-to-0.8.1
    resources:
    - name: pool-gdkl
      namespace: openebs
      kind: CStorPool
      apiVersion: openebs.io/v1alpha1
    - name: pool-tdsv
      namespace: openebs
      kind: CStorPool
      apiVersion: openebs.io/v1alpha1
    - name: pool-ddcm
      namespace: openebs
      kind: CStorPool
      apiVersion: openebs.io/v1alpha1
    data:
    - name: baseVersion
      value: 0.8.0
    - name: targetVersion
      value: 0.8.1
    - name: retryOnError
      value: true
    - name: rollbackOnError
      value: false
```
UpgradeResult custom resource will embed the provided config details in its config section:
```yaml
kind: UpgradeResult
metadata:
  name: <job-name>-<resource-name>
  annotations:
    ownerReferences:
      kind: Job
      name: <name of job>
      uuid: <uuid of job>
    openebs.io/upgrade-config: <name of configmap>
    openebs.io/cas-template: <name of castemplate>
spec:
  config:
    name: pool-gdkl
    namespace: openebs
    kind: CStorPool
    apiVersion: openebs.io/v1alpha1
    data:
    - name: baseVersion
      value: 0.8.0
    - name: targetVersion
      value: 0.8.1
    - name: retryOnError
      value: true
    - name: rollbackOnError
      value: false
```

### CODE REVIEW

File: pkg/executorconfig/v1alpha1/executor_config.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+// NewConfig returns empty struct of ExecutorConfig type
+func NewConfig() *ExecutorConfig {
+	ec := &ExecutorConfig{}
+	return ec
+}
+
+// WithBaseVersion adds baseVersion in ExecutorConfig object and returns updated ExecutorConfig object
+func (ec *ExecutorConfig) WithBaseVersion(baseVersion string) *ExecutorConfig {
```
Comment (imazik) [inline pkg/executorconfig/v1alpha1/executor_config.go]: Upgrade agent will use these functions to build configmap for job.

File: pkg/executorconfig/v1alpha1/executor_config.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	// use - { .TaskConfig.ObjectKind } to get this value in runtask
+	ObjectKind string `yaml:"objectKind"`
+	// CastemplateName is cast which will be used to update this resource
+	CastemplateName string `yaml:"castemplate"`
+}
+
+// ExecutorConfig provides configuration to all the castemplate for updating mentioned
+// resources values of BaseVersion, TargetVersion, UpgradeTaskCRName, RetryOnError
+// RollbackOnError passed to cast at runtime.
+type ExecutorConfig struct {
```
Comment (AmitKumarDas) [inline pkg/executorconfig/v1alpha1/executor_config.go]: The struct definitions should be in `pkg/apis/openebs.io/upgrade/v1alpha1`

File: pkg/executorconfig/v1alpha1/executor_config.go (hunk 1 above)
Comment (AmitKumarDas) [inline pkg/executorconfig/v1alpha1/executor_config.go]: Use these as methods of builder instance

```go
// pkg/upgrade/v1alpha1

import (
  apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
)

type upgrade struct {
  object *apis.Upgrade
}

type buildOption func(*upgrade)

func WithUpgradeObject(api *apis.Upgrade) buildOpt {
  return func(u *upgrade) {
    u.object = api
  }
}

func New(opts ...buildOpt) *upgrade {
  u := &upgrade{}
  for _, o := range opts {
    u(o)
  }
  return u
}

type builder struct {
  upgrade *upgrade
}

func Builder() *builder {
  return &builder{upgrade: &apis.Upgrade{}}
}

func (b *builder) WithBaseVersion(baseVer string) *builder {
  // ...
}
```

Comment (AmitKumarDas) [reply pkg/executorconfig/v1alpha1/executor_config.go]: Caller code need to decide if New() should be invoked versus Builder().
Mostly when a configmap's data is unmarshalled, then entire api object is populated. So Builder is not required in this case.

File: pkg/executorconfig/v1alpha1/executor_template_engine.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	castkey "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	cast "github.com/openebs/maya/pkg/engine"
+)
+
+type ExecutorEngine struct {
+	engine        cast.Interface // generic CAS template engine
+	defaultConfig []apis.Config  // default cas storagepool config found in CASTemplate
+	openebsConfig []apis.Config  // openebsConfig is the config that is provided
```
Comment (AmitKumarDas) [inline pkg/executorconfig/v1alpha1/executor_template_engine.go]: can we make use of Upgrade i.e. ExecutorConfig instead of apis.Config. In other words, let us pass this Upgrade to castemplate & let it handle accordingly.
We can also think of populating the Upgrade config with exact information e.g. pools, volumes, etc before passing it to castemplate.

File: buildscripts/task-executor/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,19 @@
+FROM ubuntu:18.04
```
[the remark this answers is not in the document]
Comment (imazik) [reply buildscripts/task-executor/Dockerfile]: will do it in next Pr.

File: cmd/task-executor/ExecutorJob.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: task-executor-config
+  namespace: default
+data:
+  executor.config: |
```
[the remark this answers is not in the document]
Comment (imazik) [reply cmd/task-executor/ExecutorJob.yaml]: updated

File: pkg/executorconfig/v1alpha1/executor_template_engine.go (hunk 3 above)
Comment (imazik) [reply pkg/executorconfig/v1alpha1/executor_template_engine.go]: will do in next pr.

File: pkg/executorconfig/v1alpha1/executor_config.go (hunk 1 above)
Comment (imazik) [reply pkg/executorconfig/v1alpha1/executor_config.go]: will do in next pr.

File: pkg/executorconfig/v1alpha1/executor_config.go (hunk 2 above)
Comment (imazik) [reply pkg/executorconfig/v1alpha1/executor_config.go]: will do in next pr.

File: pkg/apis/openebs.io/v1alpha1/cas_template_keys.go [hunk 6]
```diff
@@ -312,3 +323,33 @@ const (
 	// Above templating will give the current repeat resource name
 	CurrentRepeatResourceLITP ListItemsTLPProperty = "currentRepeatResource"
 )
+
+// TaskConfigTLPProperty is used to define properties that comes after TaskConfigTLP
+type TaskConfigTLPProperty string
```
[the remark this answers is not in the document]
Comment (imazik) [reply pkg/apis/openebs.io/v1alpha1/cas_template_keys.go]: will do in next pr.

File: pkg/apis/openebs.io/v1alpha1/cas_template_keys.go [hunk 7]
```diff
@@ -103,6 +103,17 @@ const (
...
+	// TaskConfigTLP is a top level property supported by CAS template engine
+	//
+	// The properties provided by the caller are placed with TaskConfigTLP
+	// as the top level property
+	//
+	// NOTE:
+	//  CAS template engine cannot modify these properties. These are the
+	// runtime properties that are provided as inputs to CAS template
+	// engine.
+	TaskConfigTLP TopLevelProperty = "TaskConfig"
```
[the remark this answers is not in the document]
Comment (imazik) [reply pkg/apis/openebs.io/v1alpha1/cas_template_keys.go]: will do in next pr.

File: cmd/task-executor/ExecutorJob.yaml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    baseVersion: v1
+    targetVersion: v2
+    upgradeTaskName: cr-name
+    retryOnError: true
+    rollbackOnError: false
+    resources:
+    - objectName: pool-123
+      objectNamespace: namespace1
+      objectKind: pool
+      CastemplateName: demo-template
```
[the remark this answers is not in the document]
Comment (imazik) [reply cmd/task-executor/ExecutorJob.yaml]: updated

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+type ExecutorConfig struct {
+	Versions
+	Options
+	// Castemplate contain castemplate name which task executor
+	// will use to upgrade single unit of resources.
+	Castemplate string `json:"Castemplate"`
+	// RuntimeConfig is used to provide some runtime config to
+	// castemplate engine. Task executor will directly copy this
+	// config to castemplate engine.
+	RuntimeConfig apis.Config `json:"runtimeConfig"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: let us avoid depending on apis.Config.
Create a well defined structure here.
Transform this to cast config in the business logic.

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
+/*
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: can we rename the file to config.go
we understand that this is related to upgrade.

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+)
+
+// ExecutorConfig represents task executor config using these config
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: I felt avoiding who will use this schema might be good.
Can we rename this to UpgradeConfig.
In our case, it used by job or task executor or something later on.

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+// task executor runs castemplate engine for each given resources.
+// This contains Versions, upgrade Options and castemplate name.
+type ExecutorConfig struct {
+	Versions
+	Options
+	// Castemplate contain castemplate name which task executor
+	// will use to upgrade single unit of resources.
+	Castemplate string `json:"Castemplate"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: Can we make this camelCased i.e. `casTemplate`?

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 13]
```diff
@@ -45,8 +46,19 @@ type UpgradeResult struct {
 
 // UpgradeResultConfig represents the entire config
 // of UpgradeResult i.e. same as task-executor job config
+// with single unit of upgrade in resource details. pv or sp
+// will be single unit of upgrade.
 type UpgradeResultConfig struct {
-	// TODO: Add Task-executor job config here
+	ResourceDetails
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Can we avoid this structure completely and make use of ExecutorConfig i.e. UpgradeConfig

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go (hunk 10 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: done.

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go (hunk 12 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: done

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go (hunk 9 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: added.

File: pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go (hunk 11 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/taskconfig.go]: done. please check.

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+package v1alpha1
+
+// Config represents configuration for a job or executor or some other
+// coomponent. This contains Resources, JobId, Castemplate and RuntimeConfigs
+// Mentioned Resources will be updated using mention castemplate
+type Config struct {
+	// Castemplate contain castemplate name which task executor
+	// will use to upgrade single unit of resources.
+	Castemplate string `json:"casTemplate"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]:
```go
CASTemplate string `json:"casTemplate"`
```

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+// Config represents configuration for a job or executor or some other
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: spellchecks !!

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+// coomponent. This contains Resources, JobId, Castemplate and RuntimeConfigs
+// Mentioned Resources will be updated using mention castemplate
+type Config struct {
+	// Castemplate contain castemplate name which task executor
+	// will use to upgrade single unit of resources.
+	Castemplate string `json:"casTemplate"`
+	// RuntimeConfigs is used to provide some runtime config to
+	// castemplate engine. Task executor will directly copy this
+	// config to castemplate engine.
+	RuntimeConfigs []RuntimeConfig `json:"runtimeConfig"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: Can we rename this to:
```go
Data []RuntimeConfig `json:"data"`
```

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+	// JobId contains unique id for each job
+	JobId string `json:"jobId"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: Why is this required.
A job will make use of the Config is what i understand.
So the job specs or the job code will make use of the ConfigMap

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+// RuntimeConfig holds a runtime configuration for executor
+type RuntimeConfig struct {
+	// Name of the config
+	Name string `json:"name"`
+	// Value represents any specific value that is applicable
+	// to this config
+	Value string `json:"value"`
+	// Data represents an arbitrary map of key value pairs
+	Data map[string]string `json:"data"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: Can we rename this to:
```go
Entries map[string]string `json:"entries"`
```

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+	// config to castemplate engine.
+	RuntimeConfigs []RuntimeConfig `json:"runtimeConfig"`
+	// JobId contains unique id for each job
+	JobId string `json:"jobId"`
+	// Resources contains list of resources which we are going to upgrade
+	Resources []ResourceDetails `json:"Resources"`
+}
+
+// RuntimeConfig holds a runtime configuration for executor
+type RuntimeConfig struct {
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: Can we rename this to `DataItem`

So `UpgradeConfig uses Data uses []DataItem`

IMO avoiding word `runtime` will be good.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 20]
```diff
@@ -36,17 +36,22 @@ type UpgradeResult struct {
...
 }
 
-// UpgradeResultConfig represents the entire config
-// of UpgradeResult i.e. same as task-executor job config
-type UpgradeResultConfig struct {
-	// TODO: Add Task-executor job config here
+// UpgradeConfig represents the config of UpgradeResult i.e.
+// It contains resource details of single unit of upgrade and
+// all runtime configuration.
+type UpgradeConfig struct {
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Can we rename this to `UpgradeResultConfig`

It will be in the same lines as `UpgradeResultTask` & `UpgradeResultStatus`

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+// Config represents configuration for a job or executor or some other
+// coomponent. This contains Resources, JobId, Castemplate and RuntimeConfigs
+// Mentioned Resources will be updated using mention castemplate
+type Config struct {
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: Can we rename this to `UpgradeConfig`

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 22]
```diff
@@ -36,17 +36,22 @@ type UpgradeResult struct {
...
+	ResourceDetails
+	// RuntimeConfigs is used to provide some runtime config to
+	// castemplate engine. Task executor will directly copy this
+	// config to castemplate engine.
+	RuntimeConfigs []RuntimeConfig `json:"runtimeConfigs"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Can we rename this to:
```go
Data []DataItem `json:"data"`
```

IMO this is the merged info from UpgradeConfig.Data & CASTemplate.Defaults

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 23]
```diff
@@ -130,6 +135,8 @@ type ResourceDetails struct {
 	APIVersion string `json:"apiVersion"`
 	// Namespace of the resource
 	Namespace string `json:"namespace"`
+	// Generation of resource
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: The comment is not clear. Can we explain further.

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go (hunk 14 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: done

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go (hunk 16 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: sound good. done.

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go (hunk 17 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: I thought this will helpful if we want to list out result cr created by one particular job.
As we will set Owner Reference in  annotations this not required.
removed it.

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go (hunk 18 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: sounds good done

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go (hunk 19 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: done

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go (hunk 20 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Changed. I thought other way that's why I changed both config names.

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go (hunk 21 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: Changed. I thought other way that's why I changed both config names.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go (hunk 22 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: done

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+type UpgradeConfig struct {
+	// CASTemplate contain castemplate name which task executor
+	// will use to upgrade single unit of resources.
+	CASTemplate string `json:"casTemplate"`
+	// Data is used to provide some runtime configurations to
+	// castemplate engine. Task executor will directly copy these
+	// configurations to castemplate engine.
+	Data []DataItem `json:"data"`
+	// Resources contains list of resources which we are going to upgrade
+	Resources []ResourceDetails `json:"Resources"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: json name should be camelCased. In this case it should be `json: "resources"`

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go (hunk 23 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: done

File: pkg/apis/openebs.io/upgrade/v1alpha1/config.go (hunk 24 above)
Comment (imazik) [reply pkg/apis/openebs.io/upgrade/v1alpha1/config.go]: done.
